fvm/transactionVerifier.go

```diff
@@ -141,73 +237,121 @@ func (v *TransactionVerifier) verifyTransaction(
-
-		accountKey, err := accounts.GetPublicKey(txSig.Address, txSig.KeyIndex)
+) error {
+	foundProposalSignature := false
+	for _, signature := range signatures {
+		accountKey, err := accounts.GetPublicKey(
+			signature.Address,
+			signature.KeyIndex)
 		if err != nil {
-			return nil, false, errorBuilder(txSig, err)
+			return signature.newError(err)
 		}
-		err = v.verifyAccountSignature(accountKey, txSig, message, errorBuilder)
-		if err != nil {
-			return nil, false, err
+
+		if accountKey.Revoked {
+			return signature.newError(
+				fmt.Errorf("account key has been revoked"))
 		}
-		if !proposalKeyVerified && v.sigIsForProposalKey(txSig, proposalKey) {
-			proposalKeyVerified = true
+
+		signature.accountKey = accountKey
+
+		if !foundProposalSignature && signature.matches(proposalKey) {
+			foundProposalSignature = true
 		}
+	}
 
-		weights[txSig.Address] += accountKey.Weight
+	if !foundProposalSignature {
+		return fmt.Errorf(
+			"either the payload or the envelope should provide proposal " +
+				"signatures")
 	}
 
-	return
+	return nil
 }
 
-// verifyAccountSignature verifies that an account signature is valid for the
-// account and given message.
-//
-// If the signature is valid, this function returns the associated account key.
-//
-// An error is returned if the account does not contain a public key that
-// correctly verifies the signature against the given message.
-func (v *TransactionVerifier) verifyAccountSignature(
-	accountKey flow.AccountPublicKey,
-	txSig flow.TransactionSignature,
-	message []byte,
-	errorBuilder func(flow.TransactionSignature, error) errors.CodedError,
+func (v *TransactionVerifier) verifyAccountSignatures(
+	signatures []*signatureEntry,
 ) error {
+	toVerifyChan := make(chan *signatureEntry, len(signatures))
+	verifiedChan := make(chan *signatureEntry, len(signatures))
 
-	if accountKey.Revoked {
-		return errorBuilder(txSig, fmt.Errorf("account key has been revoked"))
+	verificationConcurrency := v.VerificationConcurrency
+	if len(signatures) < verificationConcurrency {
+		verificationConcurrency = len(signatures)
 	}
 
-	valid, err := crypto.VerifySignatureFromTransaction(
-		txSig.Signature,
-		message,
-		accountKey.PublicKey,
-		accountKey.HashAlgo,
-	)
-	if err != nil {
-		return errorBuilder(txSig, err)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	wg := sync.WaitGroup{}
+	wg.Add(verificationConcurrency)
+
+	for i := 0; i < verificationConcurrency; i++ {
+		go func() {
+			defer wg.Done()
+
+			for entry := range toVerifyChan {
+				err := entry.verify()
+
+				verifiedChan <- entry
+
+				if err != nil {
+					// Signal to other workers to early exit
+					cancel()
+					return
+				}
+
+				select {
+				case <-ctx.Done():
+					// Another worker has error-ed out.
+					return
+				default:
+					// continue
+				}
+			}
+		}()
 	}
 
-	if valid {
+	for _, entry := range signatures {
+		toVerifyChan <- entry
+	}
+	close(toVerifyChan)
```

Comment:
nit: you may want to move this to the top of the function.

Reply:
your suggestion serializes queuing ...

in the current form, the goroutines could start processing in the background while the main function is queuing up the entries

your proposal is to queue up all the entires first, then start the goroutines